NETTY-7 use exception handlers for streaming methods too

albertshau opened the pull request:
Streaming methods (those that return BodyConsumer) were not using
the exception handler. This fix will catch any throwable from
invoking the method that returns the BodyConsumer, as well as any
throwables from chunk() and finished() calls on the BodyConsumer
itself, and send the throwable to the ExceptionHandler.

albertshau committed:
NETTY-7 use exception handlers for streaming methods too

Streaming methods (those that return BodyConsumer) were not using
the exception handler. This fix will catch any throwable from
invoking the method that returns the BodyConsumer, as well as any
throwables from chunk() and finished() calls on the BodyConsumer
itself, and send the throwable to the ExceptionHandler.

poornachandra commented on src/main/java/co/cask/http/HttpMethodInfo.java:
```diff
@@ -114,7 +115,8 @@ private void bodyConsumerChunk(ChannelBuffer buffer) throws HandlerException {
       bodyConsumer.chunk(buffer, responder);
     } catch (Throwable t) {
-      throw bodyConsumerError(t);
+      bodyConsumerError(t);
```
It would be a good idea to wrap invocation of `consumer.handleError()` in a try blcok.

albertshau replied:
yeah i guess we should do that, though its a bit weird on which exception the exception handler should get in that case. I guess it should get the handleError exception

poornachandra commented on src/main/java/co/cask/http/HttpMethodInfo.java:
```diff
@@ -125,19 +127,21 @@ private void bodyConsumerChunk(ChannelBuffer buffer) throws HandlerException {
+  private void bodyConsumerError(Throwable cause) {
     BodyConsumer consumer = bodyConsumer;
     bodyConsumer = null;
     consumer.handleError(cause);
```
It would be a good idea to wrap invocation of consumer.handleError() in a try block.

albertshau committed:
NETTY-7 handle exception for handleError as well

Since handleError is implemented by users, wrap it in a try catch
as well and use the exceptionHandler on whatever it threw.

albertshau committed:
NETTY-7 add javadoc to clarify intended behavior for handleError

poornachandra commented on src/main/java/co/cask/http/HttpMethodInfo.java:
```diff
@@ -114,7 +115,13 @@ private void bodyConsumerChunk(ChannelBuffer buffer) throws HandlerException {
+      try {
+        bodyConsumerError(t);
+      } catch (Throwable t2) {
+        exceptionHandler.handle(t2, request, responder);
```
It would be a good idea to log the first exception, so that we don't lose it in this case.

albertshau committed:
NETTY-7 log original throwable so its not lost.

poornachandra commented:
LGTM
